examples/native-fungible/src/contract.rs

```diff
@@ -0,0 +1,287 @@
+
+                system_api::claim(source_account, target_account, amount);
+
+                if fungible_source_account.chain_id == system_api::current_chain_id() {
+                    Ok(self.get_transfer_outcome(
+                        fungible_source_account.owner,
+                        fungible_target_account,
+                        amount,
+                    ))
+                } else {
+                    // If different chain, send message that will be ignored so the app gets auto-deployed
+                    let message = Message::Withdraw {
+                        owner: fungible_source_account.owner,
+                        amount,
+                        target_account: fungible_target_account,
+                    };
+                    Ok(ExecutionOutcome::default()
+                        .with_authenticated_message(fungible_source_account.chain_id, message))
+                }
+            }
+        }
+    }
+
+    async fn execute_message(
+        &mut self,
+        context: &MessageContext,
+        message: Self::Message,
+    ) -> Result<ExecutionOutcome<Self::Message>, Self::Error> {
+        // Messages for now don't do anything, just pass messages around
+        match message {
+            Message::Credit {
+                amount: _,
+                target: _,
+                source: _,
+            } => Ok(ExecutionOutcome::default()),
+            Message::Withdraw {
+                owner,
+                amount,
+                target_account,
+            } => {
+                Self::check_account_authentication(context.authenticated_signer, owner)?;
+                Ok(self.get_transfer_outcome(owner, target_account, amount))
+            }
+        }
+    }
+
+    async fn handle_application_call(
+        &mut self,
+        context: &CalleeContext,
+        call: ApplicationCall,
+        _forwarded_sessions: Vec<SessionId>,
+    ) -> Result<
+        ApplicationCallOutcome<Self::Message, Self::Response, Self::SessionState>,
+        Self::Error,
+    > {
+        match call {
+            ApplicationCall::Balance { owner } => {
+                let owner = self.normalize_owner(owner);
+
+                let mut outcome = ApplicationCallOutcome::default();
+                let balance = system_api::current_owner_balance(owner);
+                outcome.value = FungibleResponse::Balance(balance);
+                Ok(outcome)
+            }
+
+            ApplicationCall::Transfer {
+                owner,
+                amount,
+                destination,
+            } => {
+                Self::check_account_authentication(context.authenticated_signer, owner)?;
+                let owner = self.normalize_owner(owner);
+                let target_account = self.destination_to_account(destination);
+
+                system_api::transfer(Some(owner), target_account, amount);
+                Ok(ApplicationCallOutcome::default())
```

Comment:
Shouldn't we also `get_transfer_outcome` here, and include the `Credit` message if it's for another chain?